File: 백준/Silver/11724. 연결 요소의 개수/연결 요소의 개수.py
import sys
sys.setrecursionlimit(10**6)
def DFS(n):
    visited[n] = 1
    for j in arr[n]:
        if visited[j] == 0:
            DFS(j)

N,M = map(int,sys.stdin.readline().split())
arr = [[] for _ in range(N+1)]
for _ in range(M):
    u,v = map(int,sys.stdin.readline().split())
    arr[u].append(v)
    arr[v].append(u)
count = 0
visited = [0] * (N+1)

for i in range(1,N+1):
    if visited[i] == 0:
        if not arr[i]:
            count += 1
            visited[i] = 1
        else:
            count +=1
            DFS(i)
print(count)

File: 백준/Silver/11724. 연결 요소의 개수/연결 요소의 개수.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;
import java.util.StringTokenizer;

public class Main {
    static int[][] graph;
    static int count = 0;
    static boolean[] visited;
    static Stack<Integer> stack;
    public static void dfs(int n){
        stack.push(n);
        while(!stack.isEmpty()){
            int node = stack.pop();
            if(!visited[node]){
                visited[node] = true;
            }
            for(int i = 1; i<graph.length;i++){
                if (graph[node][i] == 1 && !visited[i]){
                    visited[i] = true;
                    stack.push(i);
                }
            }
        }
    }
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());
        graph = new int[N+1][N+1];
        for(int i = 0; i<M; i++){
            StringTokenizer s2 = new StringTokenizer(br.readLine());
            int u = Integer.parseInt(s2.nextToken());
            int v= Integer.parseInt(s2.nextToken());
            graph[u][v] = 1;
            graph[v][u] = 1;
        }
        stack = new Stack<>();
        visited = new boolean[N+1];
        for(int i = 1; i<=N; i++){
            if(!visited[i]){
                dfs(i);
                count++;
            }
        }
        System.out.println(count);

    }
}
